src/main/java/com/iota/iri/LedgerValidator.java

```diff
@@ -198,103 +184,90 @@ private void updateConsistentHashes(final Snapshot snapshot, Hash tip, int index
                         stateDiffViewModel.store(tangle);
                     }
-                    latestSnapshot.merge(latestSnapshot.patch(stateDiffViewModel.getDiff(), milestone.index()));
+                    milestone.latestSnapshot.apply(currentState, milestoneVM.index());
                 }
             }
             return hasSnapshot;
+        } finally {
+            milestone.latestSnapshot.rwlock.writeLock().unlock();
         }
     }
 
-    public boolean checkConsistency(final Snapshot snapshotReference, List<Hash> hashes) throws Exception {
-        Snapshot snapshot;
-        synchronized (snapshotReference.snapshotSyncObject) {
-            snapshot = new Snapshot(snapshotReference);
-        }
-        for(Hash hash: hashes) {
-            if (!isTipConsistent(snapshot, hash)) return false;
+    public boolean checkConsistency(List<Hash> hashes) throws Exception {
+        Set<Hash> visitedHashes = new HashSet<>();
+        Map<Hash, Long> diff = new HashMap<>();
+        for (Hash hash : hashes) {
+            if (!updateDiff(visitedHashes, diff, hash)) return false;
         }
         return true;
     }
 
-    public boolean isTipConsistent(Snapshot snapshot, Hash tip) throws Exception {
-        TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, tip);
-        if(!transactionViewModel.isSolid()) {
+    public boolean updateDiff(Set<Hash> approvedHashes, final Map<Hash, Long> diff, Hash tip) throws Exception {
+        if(!TransactionViewModel.fromHash(tangle, tip).isSolid()) {
             return false;
         }
-        boolean isConsistent;
-        synchronized (snapshot.snapshotSyncObject) {
-            synchronized (snapshot.approvalsSyncObject) {
-                if (!(isConsistent = snapshot.approvedHashes.contains(tip))) {
-                    Hash tail = transactionViewModel.getHash();
-                    int latestSyncIndex = snapshot.index();
-                    Map<Hash, Long> currentState = getLatestDiff(snapshot, tail, latestSyncIndex, false);
-                    isConsistent = currentState != null && snapshot.patch(currentState, latestSyncIndex).isConsistent();
-                    if (isConsistent) {
-                        updateConsistentHashes(snapshot, tip, latestSyncIndex);
-                        snapshot.merge(snapshot.patch(currentState, latestSyncIndex));
-                    }
-                }
-            }
+        if (approvedHashes.contains(tip)) return true;
+        Set<Hash> visitedHashes = new HashSet<>(approvedHashes);
+        Map<Hash, Long> currentState = getLatestDiff(visitedHashes, tip, milestone.latestSnapshot.index(), false);
+        if (currentState == null) return false;
+        boolean isConsistent = Snapshot.isConsistent(milestone.latestSnapshot.patchedDiff(currentState));
+        if (isConsistent) {
+            currentState.forEach((key, value) -> {
+                diff.computeIfPresent(key, ((hash, aLong) -> value + aLong));
```

Comment:
This will return none if key is absent. Check for that and only all `putIfAbsent` conditionally?